keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -356,28 +289,38 @@ def generate(
         """
         if self.preprocessor is None:
             raise ValueError(
-                "`self.preprocessor` is None, please make sure `preprocessor` "
-                "is set before calling `generate`."
+                "`self.preprocessor` is `None`, please make sure "
+                "`preprocessor` is set before calling `generate`."
             )
-        end_token_id = self.preprocessor.tokenizer.end_token_id
-
         sampler = keras_nlp.samplers.get(sampler)
-        if hasattr(self, "jit_compile"):
-            # `jit_compile` is a public property as of tf 2.12. hasattr is for
-            # backward compat.
-            sampler.jit_compile = self.jit_compile
+        # `jit_compile` is a property of keras.Model after tf 2.12.
+        # Use `getattr()` for backwards compatibility.
+        sampler.jit_compile = getattr(self, "jit_compile", True)
         sampler.run_eagerly = self.run_eagerly
-        x, _, _ = self.preprocessor(prompt)
-        if len(x["token_ids"].shape) == 1:
-            x["token_ids"] = x["token_ids"][tf.newaxis, :]
-            x["padding_mask"] = x["padding_mask"][tf.newaxis, :]
-        _, cache = self.build_initial_cache(x, max_length)
-        next_token_probability = self._get_token_probability
+
+        # Tokenize.
+        prompt = self.preprocessor.tokenizer(prompt)
+
+        # Create and seed the cache before generation.
+        token_ids = prompt
+        if prompt.shape.rank == 1:
+            token_ids = tf.RaggedTensor.from_tensor(prompt[tf.newaxis, :])
+        token_ids = token_ids.to_tensor(shape=(None, max_length))
+        # Pass a padding mask of all ones when seeing the cache. The mask will
+        # not affect cached key/values for input tokens we care about.
+        padding_mask = tf.ones_like(token_ids, dtype=tf.bool)
+        batch_size = tf.shape(token_ids)[0]
+        cache = self.build_empty_cache(batch_size, max_length)
+        _, cache = self.call_with_cache(token_ids, padding_mask, cache, 0)
```

Comment:
I am thinking we can fold this initial call into the `sampler` call. When `cache_index=0`, the code knows there is no cached computed before. But let's discuss it in your one pager and launch this PR!

Reply:
I think we should move compilation up the model for a whole host of reasons--not insignificantly including consistency with core Keras in terms of how it handles compilation. But agreed let's keep discussing this as a follow up!